src/geometry/euclidean.lean

```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+  { intro h,
+    rcases h with ⟨hx, ⟨r, ⟨hr, hy⟩⟩⟩,
+    rw hy,
+    exact abs_inner_div_norm_mul_norm_eq_one_of_ne_zero_of_ne_zero_mul hx hr }
+end
+
+/-- The inner product of two vectors, divided by the product of their
+norms, has value 1 if and only if they are nonzero and one is
+a positive multiple of the other. -/
+lemma inner_div_norm_mul_norm_eq_one_iff (x y : V) :
+  inner x y / (∥x∥ * ∥y∥) = 1 ↔ (x ≠ 0 ∧ ∃ (r : ℝ), 0 < r ∧ y = r • x) :=
+begin
+  split,
+  { intro h,
+    have ha := h,
+    apply_fun abs at ha,
+    norm_num at ha,
+    rcases (abs_inner_div_norm_mul_norm_eq_one_iff x y).1 ha with ⟨hx, ⟨r, ⟨hr, hy⟩⟩⟩,
+    use [hx, r],
+    refine and.intro _ hy,
+    by_contradiction hrneg,
+    rw hy at h,
+    rw inner_div_norm_mul_norm_eq_neg_one_of_ne_zero_of_neg_mul hx
+      (lt_of_le_of_ne' (le_of_not_lt hrneg) hr) at h,
+    norm_num at h },
+  { intro h,
+    rcases h with ⟨hx, ⟨r, ⟨hr, hy⟩⟩⟩,
+    rw hy,
+    exact inner_div_norm_mul_norm_eq_one_of_ne_zero_of_pos_mul hx hr }
+end
+
+/-- The inner product of two vectors, divided by the product of their
+norms, has value -1 if and only if they are nonzero and one is
+a negative multiple of the other. -/
+lemma inner_div_norm_mul_norm_eq_neg_one_iff (x y : V) :
+  inner x y / (∥x∥ * ∥y∥) = -1 ↔ (x ≠ 0 ∧ ∃ (r : ℝ), r < 0 ∧ y = r • x) :=
+begin
+  split,
+  { intro h,
+    have ha := h,
+    apply_fun abs at ha,
+    norm_num at ha,
+    rcases (abs_inner_div_norm_mul_norm_eq_one_iff x y).1 ha with ⟨hx, ⟨r, ⟨hr, hy⟩⟩⟩,
+    use [hx, r],
+    refine and.intro _ hy,
+    by_contradiction hrpos,
+    rw hy at h,
+    rw inner_div_norm_mul_norm_eq_one_of_ne_zero_of_pos_mul hx
+      (lt_of_le_of_ne' (le_of_not_lt hrpos) hr.symm) at h,
+    norm_num at h },
+  { intro h,
+    rcases h with ⟨hx, ⟨r, ⟨hr, hy⟩⟩⟩,
+    rw hy,
+    exact inner_div_norm_mul_norm_eq_neg_one_of_ne_zero_of_neg_mul hx hr }
+end
+
+/-- The undirected angle between two vectors. If either vector is 0,
+this is π/2. -/
+def angle_of_vectors (x y : V) : ℝ := (inner x y / (∥x∥ * ∥y∥)).arccos
+
+/-- The angle between two vectors does not depend on their order. -/
+lemma angle_of_vectors_comm (x y : V) : angle_of_vectors x y = angle_of_vectors y x :=
+begin
+  unfold angle_of_vectors,
+  rw [inner_comm, mul_comm]
+end
+
+/-- The angle between the negation of two vectors. -/
+@[simp] lemma angle_of_vectors_of_neg_of_neg (x y : V) :
+  angle_of_vectors (-x) (-y) = angle_of_vectors x y :=
+begin
+  unfold angle_of_vectors,
+  rw [inner_neg_neg, norm_neg, norm_neg]
+end
+
+/-- The angle between two vectors is nonnegative. -/
+lemma angle_of_vectors_nonneg (x y : V) : 0 ≤ angle_of_vectors x y :=
+real.arccos_nonneg _
+
+/-- The angle between two vectors is at most π. -/
+lemma angle_of_vectors_le_pi (x y : V) : angle_of_vectors x y ≤ real.pi :=
+real.arccos_le_pi _
+
+/-- The angle between a vector and the negation of another vector. -/
+lemma angle_of_vectors_of_neg_right (x y : V) :
```

Comment:
`of_neg_right` would mean that "neg_right" is the description of an assumption. So this could become
`angle_neg_right`
Similar comment applies elsewhere.

Reply:
Renamed various such lemmas.